cmd/conversion/main.go

```diff
@@ -0,0 +1,330 @@
+	return &TreeNode{
+		ClassName:  className,
+		Attributes: attributes,
+		Children:   make(map[string]*TreeNode),
+	}
+}
+
+func constructTree(resources []map[string]interface{}) map[string]interface{} {
+	rootNode := NewTreeNode("uni", map[string]interface{}{"dn": "uni"})
+
+	dnMap := make(map[string]*TreeNode)
+	dnMap["uni"] = rootNode
+
+	for _, resourceList := range resources {
+		for resourceType, resourceData := range resourceList {
+			resourceAttributes := resourceData.(map[string]interface{})
+			attributes := safeMapInterface(resourceAttributes, "attributes")
+			dn := safeString(attributes, "dn")
+
+			var children []map[string]interface{}
+			if rawChildren, ok := resourceAttributes["children"].([]interface{}); ok {
+				for _, child := range rawChildren {
+					if childMap, ok := child.(map[string]interface{}); ok {
+						children = append(children, childMap)
+					}
+				}
+			}
+
+			buildParentPath(dnMap, rootNode, resourceType, dn, attributes, children)
+		}
+	}
+
+	return map[string]interface{}{rootNode.ClassName: exportTree(rootNode)}
+}
+
+func buildParentPath(dnMap map[string]*TreeNode, rootNode *TreeNode, resourceType, dn string, attributes map[string]interface{}, children []map[string]interface{}) {
+	if dn == "" && resourceType == "" {
+		return
+	}
+
+	cursor := rootNode
+	if dn != "" {
+		cursor = traverseOrCreatePath(dnMap, rootNode, resourceType, dn)
+	}
+
+	var leafNode *TreeNode
+	if existingLeafNode, exists := dnMap[dn]; exists {
+
+		for key, value := range attributes {
+			existingLeafNode.Attributes[key] = value
+		}
+		leafNode = existingLeafNode
+	} else {
+		leafNode = NewTreeNode(resourceType, attributes)
+		cursor.Children[dn] = leafNode
+		dnMap[dn] = leafNode
+	}
+
+	for _, child := range children {
+		for childClassName, childData := range child {
+			childAttributes := safeMapInterface(childData.(map[string]interface{}), "attributes")
+			childDn := safeString(childAttributes, "dn")
+
+			childKey := childDn
+			if childDn == "" {
+				childKey = generateUniqueKeyForNonDnNode(childClassName, childAttributes)
+			}
+
+			if _, exists := leafNode.Children[childKey]; !exists {
+				childNode := NewTreeNode(childClassName, childAttributes)
+				leafNode.Children[childKey] = childNode
+				dnMap[childKey] = childNode
+			}
+
+			if grandChildren, ok := childData.(map[string]interface{})["children"].([]interface{}); ok {
+				for _, grandchild := range grandChildren {
+					if grandchildMap, ok := grandchild.(map[string]interface{}); ok {
+						buildParentPath(dnMap, leafNode, childClassName, childDn, safeMapInterface(grandchildMap, "attributes"), []map[string]interface{}{grandchildMap})
+					}
+				}
+			}
+		}
+	}
+}
+
+func generateUniqueKeyForNonDnNode(resourceType string, attributes map[string]interface{}) string {
+	return fmt.Sprintf("%s-%v", resourceType, attributes["name"])
+}
+
+func traverseOrCreatePath(dnMap map[string]*TreeNode, rootNode *TreeNode, resourceType, dn string) *TreeNode {
```

Comment:
What is the function actually doing? Bit confused on what it is you are trying to achieve here. For more complex functions it might be an idea to add some comments in the code to make it easier to understand.